src/mockgyver.erl

```diff
@@ -1372,73 +1372,47 @@ ensure_renamed_mod_to_load_aux(RenamedMod, #modinfo{code=Code}) ->
-                                        erl_syntax:list(Args)])]),
-             [erl_syntax:clause([erl_syntax:atom(undefined)],
-                                none,
-                                [mk_call(RenamedMod, F, Args)]),
-              erl_syntax:clause([erl_syntax:variable('ActionFun')],
-                                none,
-                                [mk_call('ActionFun', Args)])])],
-    erl_syntax:function(
-      erl_syntax:abstract(F),
-      [erl_syntax:clause(Args, none, Body)]).
+    FmtNoAction =
+        fun(FnName, Args) ->
+                f("apply(~p, ~s, ~s)", [RenamedMod, FnName, Args])
+        end,
+    mk_mod(Mod, [mk_handle_undefined_function(Mod, ExportedFAs, FmtNoAction)]).
 
 mk_new_mod(Mod, ExportedFAs) ->
-    mk_mod(Mod, mk_new_funcs(Mod, ExportedFAs)).
-
-mk_new_funcs(Mod, ExportedFAs) ->
-    lists:map(fun(ExportedFA) -> mk_new_func(Mod, ExportedFA) end,
-              ExportedFAs).
-
-mk_new_func(Mod, {F, A}) ->
-    %% Generate a function like this (mod, func and arguments vary):
+    FmtNoAction =
+        fun(FnName, Args) ->
+                f("error_handler:raise_undef_exception(~p, ~s, ~s)",
+                  [Mod, FnName, Args])
+        end,
+    mk_mod(Mod, [mk_handle_undefined_function(Mod, ExportedFAs, FmtNoAction)]).
+
+mk_handle_undefined_function(Mod, ExportedFAs, FmtNoAction) ->
+    %% Parsing the string is approx 20% slower than constructing
+    %% the syntax tree using erl_syntax calls.
+    %% The string version is easier to understand though.
     %%
-    %%     func(A2, A1) ->
-    %%         case mockgyver:reg_call_and_get_action({mod,func,[A2, A1]}) of
-    %%             undefined ->
-    %%                 erlang:error(undef); % emulate undefined function
-    %%             ActionFun ->
-    %%                 ActionFun(A2, A1)
-    %%         end.
-    Args = mk_args(A),
-    Body =[erl_syntax:case_expr(
-             mk_call(mockgyver, reg_call_and_get_action,
-                     [erl_syntax:tuple([erl_syntax:abstract(Mod),
-                                        erl_syntax:abstract(F),
-                                        erl_syntax:list(Args)])]),
-             [erl_syntax:clause([erl_syntax:atom(undefined)],
-                                none,
-                                [mk_call(erlang, error,
-                                         [erl_syntax:atom(undef)])]),
-              erl_syntax:clause([erl_syntax:variable('ActionFun')],
-                                none,
-                                [mk_call('ActionFun', Args)])])],
-    erl_syntax:function(
-      erl_syntax:abstract(F),
-      [erl_syntax:clause(Args, none, Body)]).
+    %% It is many times faster than constructing a number of functions,
+    %% each containing the inner case expression, though.
+    func_from_str_fmt(
+      "'$handle_undefined_function'(FnName, Args) ->
+           case lists:member({FnName, length(Args)}, ~p) of
```

Comment:
An option here could be to do `maps:is_key({FnName, length(Args)}, ExportedFAsMap) instead. The performance impact is most often not very big commonly, but on the other hand, the downside is pretty small too:
* Runtime cost: lists:member/2 for the last element in a list of length 5000 takes 130µs, while maps:is_key/2 takes 3µs.
* Mock setup cost: constructing the mocking module gets a little bit slower, from 90ms to 140ms approx for 5000 exported functions. It is still much faster than previously 3.3s though

Reply:
Thanks for doing the measurement, I was wondering about that.

So on average lookups are 317.5ms faster with maps (5000*(130µs-3µs)/2 = 317500µs, but 90ms-140ms = 50ms slower on construction? Approx 270ms faster with maps over lists for the big module.

If I understand your numbers, maps get better than lists at ~800 functions in a module.

Reply:
Yes, if each function in a mocked module is called exactly once. I think it is more common, especially with large mocked modules, that only a smaller subset of functions get called. I'm still struggling a bit with coming up with a situation where mocked functions get called very many times. Performance testing is not realistic in a mocked scenario, but perhaps some QuickCheck/Proper like situation could happen.

Reply:
True.

Reply:
You're right that it takes many calls and big modules for maps be faster, so keeping the lists approach. (can always be revisited if something should surface)